### PR TITLE
♻️ REFACTOR: Make content parsing format agnostic

### LINKED ISSUE #71: Support for group-tab and code-tab in Markdown
As documented in this [issue](https://github.com/executablebooks/jupyter-book/issues/787#issue-657518874), `group-tab` and `code-tab` generate `reStructuredText` when used with a `Markdown` parser. This is because rST is hard coded into their directives. Would it be possible to alter these to generate the appropriate Markup, as the basic `tab` functions?

### PR DESCRIPTION
**Pull request by foster999**
Major refactor of tabs, group-tabs and code-tabs:
* Removes hard-coded reStructuredText for group and code tabs, so that this works with other parsers (inc `MyST`)
* group-tab directive now subclasses `TabDirective` to remove duplicated run code
* likewise code-tab now subclasses group-tab
* `TabDirective` and `TabsDirective` now subclass `SphinxDirective` for easier access to directive `env`

New features:
* Can now pass code-tabs with a second argument (allowing whitespace) to provide an alternative tab label
* code-tabs can now use custom lexers, which are added to the sphinx app in conf.py

Fixes #71

### CODE REVIEW

**Comment by foster999** [comment]
May need further work to meet tests

**Comment by chrisjsewell** [comment]
well that was quick lol

File: sphinx_tabs/tabs.py [hunk 1]
```diff
@@ -25,7 +26,11 @@
 LEXER_MAP = {}
 for lexer in get_all_lexers():
     for short_name in lexer[1]:
-        LEXER_MAP[short_name] = lexer[0]
+        if short_name == "r":
+            # S by default
+            LEXER_MAP[short_name] = "R"
```
Comment (foster999) [inline sphinx_tabs/tabs.py:31]: Will drop this, as changes below allow you to specify the `code-tab` label to override the pygments names

File: sphinx_tabs/tabs.py [hunk 2]
```diff
@@ -102,45 +107,39 @@ def run(self):
...
-        args = self.content[0].strip()
-        if args.startswith('{'):
-            try:
-                args = json.loads(args)
-                self.content.trim_start(1)
-            except ValueError:
-                args = {}
+        if hasattr(self, "tab_id"):
+            tab_id = self.tab_id
+            del self.tab_id
```
Comment (foster999) [inline sphinx_tabs/tabs.py:112]: Looking back, I don't think we need to delete these attrs after use
```suggestion
```

File: sphinx_tabs/tabs.py [hunk 3]
```diff
@@ -102,45 +107,39 @@ def run(self):
...
 
         text = '\n'.join(self.content)
         node = nodes.container(text)
 
         classes = 'ui bottom attached sphinx-tab tab segment'
         node['classes'] = classes.split(' ')
-        node['classes'].extend(args.get('classes', []))
+        if hasattr(self, "tab_classes"):
+            node['classes'].extend(self.tab_classes)
+            del self.tab_classes
```
Comment (foster999) [inline sphinx_tabs/tabs.py:142]:
```suggestion
```

**Comment by foster999** [comment]
> Just wait for #76 to be merged, then we can see the damage

Sounds good to me, thanks!

**Comment by foster999** [comment]
My output HTML seems to have lost its indentation, so the XML is no longer matching. @chrisjsewell any idea where/how the indentation is determined?

**Comment by chrisjsewell** [comment]
This test suite is a legacy thing I'm afraid, I haven't touched it.
If/when I do any more work on this repo, I would look to rewrite it with similar pytest fixtures as used in myst-parser, etc

**Comment by chrisjsewell** [comment]
obviously you're welcome to give it a go yourself 😬

**Comment by foster999** [comment]
Yeah I might have a stab at moving the tests over to pytest... no promises though :sweat_smile:

My question above was actually a general sphinx one - do you know how sphinx determines HTML indentation? I presumed it would just indent with each node level, but the HTML sections from my code here have no indentation at all. I'm not sure what I've broken essentially

**Comment by chrisjsewell** [comment]
Don't know, but in myst-parser etc I use beatifulsoup to prettify the HTML before diffing with pytest-regressions, which a) standardises the HTML and b) provides better error feedback

**Comment by foster999** [comment]
Thanks, that'll be really useful! `AssertionError: XML does not match` here isn't the most informative feedback

**Comment by foster999** [comment]
Ready to go after #88

**Review by chrisjsewell — CHANGES_REQUESTED** [review]
conficts to resolve 😬

**Comment by foster999** [comment]
Sorted 😄

**Comment by foster999** [comment]
Bonus to close #34

**Review by chrisjsewell — APPROVED** [review]
Cheers! well the tests pass, and I don't see any obvious errors, so 🤞 this doesn't open up any unexpected issues

**Pull request merged**